src/analysis/normed_space/bounded_linear_maps.lean

```diff
@@ -22,15 +22,115 @@ local notation f ` →_{`:50 a `} `:0 b := filter.tendsto f (nhds a) (nhds b)
 open filter (tendsto)
 open metric
 
+structure is_bounded_linear_map {k : Type*} [normed_field k]
+  {E : Type*} [normed_space k E] {F : Type*} [normed_space k F] (L : E → F)
+  extends is_linear_map k L : Prop :=
+(bound : ∃ M, M > 0 ∧ ∀ x : E, ∥L x∥ ≤ M * ∥x∥)
+
+-- refactor: bounded linear map as a bundled structure
+structure bounded_linear_map (k : Type*) [normed_field k]
+  (E : Type*) [normed_space k E] (F : Type*) [normed_space k F]
+  extends linear_map k E F :=
+(bounded : ∃ M, ∀ x : E, ∥to_fun x∥ ≤ M * ∥x∥)
+
 variables {k : Type*} [normed_field k]
 variables {E : Type*} [normed_space k E]
 variables {F : Type*} [normed_space k F]
 variables {G : Type*} [normed_space k G]
 
-structure is_bounded_linear_map {k : Type*}
-  [normed_field k] {E : Type*} [normed_space k E] {F : Type*} [normed_space k F] (L : E → F)
-  extends is_linear_map k L : Prop :=
-(bound : ∃ M, M > 0 ∧ ∀ x : E, ∥ L x ∥ ≤ M * ∥ x ∥)
+namespace bounded_linear_map
+
+instance : has_coe (bounded_linear_map k E F) (linear_map k E F) :=
+  ⟨to_linear_map⟩
+
+instance : has_coe_to_fun (bounded_linear_map k E F) :=
+  ⟨_, λ f, f.to_fun⟩
+
+@[extensionality] theorem ext {f g : bounded_linear_map k E F}
+  (h : ∀ x, f x = g x) : f = g :=
+  by cases f; cases g; congr' 1; ext x; apply h
+
+theorem ext_iff {f g : bounded_linear_map k E F} : f = g ↔ ∀ x, f x = g x :=
+  ⟨λ h x, by rintro; rw h, ext⟩
+
+variable {f : bounded_linear_map k E F}
+
+lemma map_zero : f 0 = 0 := linear_map.map_zero _
+
+lemma has_pos_bound : ∃ M > 0, ∀ x, ∥f x∥ ≤ M * ∥x∥ :=
+  let ⟨M, hf⟩ := f.bounded in
+  ⟨max 1 M, lt_of_lt_of_le zero_lt_one (le_max_left 1 _), λ _, le_trans
+  (hf _) (mul_le_mul_of_nonneg_right (le_max_right _ _) (norm_nonneg _))⟩
+
+-- some lemmas about things having bounds and things being bounds
+lemma ratio_has_pos_bound : ∃ M > 0, ∀ x ≠ 0, ∥f x∥ / ∥x∥ ≤ M :=
```

Comment:
Isn't `x ≠ 0` a useless assumption?

Reply:
fixed in 8946ddcaee6ff2c15022befd6b1668cb11c2fca5 .